resources/images/exif/exif.go

```diff
@@ -113,127 +114,150 @@ func (d *Decoder) Decode(r io.Reader) (ex *Exif, err error) {
 		}
 	}()
 
-	var x *_exif.Exif
-	x, err = _exif.Decode(r)
+	rawExif, err := _exif.SearchAndExtractExifWithReader(r)
 	if err != nil {
-		if err.Error() == "EOF" {
-			// Found no Exif
-			return nil, nil
-		}
-		return
+		return nil, err
 	}
 
+	im, _ := exifcommon.NewIfdMappingWithStandard()
+	ti := _exif.NewTagIndex()
+
+	_, index, _ := _exif.Collect(im, ti, rawExif)
+
 	var tm time.Time
 	var lat, long float64
 
-	if !d.noDate {
-		tm, _ = x.DateTime()
+	rootIfd := index.RootIfd
+
+	ifd, err := rootIfd.ChildWithIfdPath(exifcommon.IfdGpsInfoStandardIfdIdentity)
+	if !d.noLatLong && err == nil {
+		gpsInfo, _ := ifd.GpsInfo()
+
+		lat = gpsInfo.Latitude.Decimal()
+		long = gpsInfo.Longitude.Decimal()
 	}
 
-	if !d.noLatLong {
-		lat, long, _ = x.LatLong()
+	exifIfd, _ := rootIfd.ChildWithIfdPath(exifcommon.IfdExifStandardIfdIdentity)
+
+	if results, err := exifIfd.FindTagWithId(dateTimeTagId); !d.noDate && err == nil && len(results) == 1 {
+		dateTimeTagEntry := results[0]
+		dateTimeRaw, _ := dateTimeTagEntry.Value()
+		dateTimeString := dateTimeRaw.(string)
+
+		tm, _ = time.Parse(exifTimeLayout, dateTimeString)
 	}
 
-	walker := &exifWalker{x: x, vals: make(map[string]interface{}), includeMatcher: d.includeFieldsRe, excludeMatcher: d.excludeFieldsrRe}
-	if err = x.Walk(walker); err != nil {
-		return
+	vals, err := extractTags(&index, d.includeFieldsRe, d.excludeFieldsrRe)
+	if err != nil {
+		// No Exif metadata
+		return nil, nil
 	}
 
-	ex = &Exif{Lat: lat, Long: long, Date: tm, Tags: walker.vals}
+	ex = &Exif{Lat: lat, Long: long, Date: tm, Tags: vals}
 
-	return
+	return ex, nil
 }
 
-func decodeTag(x *_exif.Exif, f _exif.FieldName, t *tiff.Tag) (interface{}, error) {
-	switch t.Format() {
-	case tiff.StringVal, tiff.UndefVal:
-		s := nullString(t.Val)
-		if strings.Contains(string(f), "DateTime") {
-			if d, err := tryParseDate(x, s); err == nil {
+// Code borrowed from exif.DateTime and adjusted.
+func tryParseDate(s string) (time.Time, error) {
+	dateStr := strings.TrimRight(s, "\x00")
+	// TODO(bep): look for timezone offset, GPS time, etc.
+	timeZone := time.Local
+	return time.ParseInLocation(exifTimeLayout, dateStr, timeZone)
+}
+
+func processRational(n int64, d int64) interface{} {
+	rat := big.NewRat(n, d)
+	if n != 1 {
```

Comment:
@bep This logic existed in the previous code, making the value of rational exif tags either `float64` or `big.Rat` type depending on the numerator. Have copied the logic as such and added corresponding tests; however, let me know if returning just one of the types (`float64` or `big.Rat`) all the time makes more sense.